### PR TITLE
feature/CIV-1343/Agreed response deadline

### PR DESCRIPTION
**Pull request by laperla11**
**Before creating a pull request make sure that:**

- [x] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [x] tests have been updated / new tests has been added (if needed)

Please remove this line and everything above and fill the following sections:

### JIRA link ###
https://tools.hmcts.net/jira/browse/CIV-1343

### Change description ###
- Allow defendant to input agreed response deadline

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

**Review by laarmada — CHANGES_REQUESTED** [review]

**Review by laarmada — CHANGES_REQUESTED** [review]

File: src/main/views/features/response/responseDeadline/agreed-response-deadline.njk [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+{% block content %}
+  {{ govukBackLink({
+    text: t('COMMON.BACK'),
+    href: backLink
+  }) }}
+  <div class="govuk-width-container">
+    <main class="govuk-main-wrapper" role="main">
+      {% set errors = form.getErrors() %}
+      {% if form.hasErrors() %}
+        {{ govukErrorSummary({
```
Comment (commitBlob) [inline src/main/views/features/response/responseDeadline/agreed-response-deadline.njk:32]: can we use errorSummary macro please

File: src/test/unit/views/features/response/responseDeadline/agreed-response-deadline.test.ts [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+      expect(errorSummary[0].innerHTML).toContain('There was a problem');
+    });
+
+    it('should display correct error summary message with correct link for agreed response date', () => {
+      const errorSummaryMessage = getErrorSummaryListElement(0);
+      expect(errorSummaryMessage.innerHTML).toContain(VALID_AGREED_RESPONSE_DATE);
+      expect(errorSummaryMessage.getElementsByTagName('a')[0].getAttribute('href'))
+        .toContain('#date');
+    });
+
```
Comment (commitBlob) [inline src/test/unit/views/features/response/responseDeadline/agreed-response-deadline.test.ts:119]: can we get rid of multiple blank lines please, one is enough

**Review by laarmada — APPROVED** [review]

File: src/main/common/form/validationErrors/errorMessageConstants.ts [hunk 3]
```diff
@@ -1,5 +1,8 @@
 export const VALID_PHONE_NUMBER = 'Enter a phone number in the correct format';
+export const DATE_NOT_MORE_THAN_28_DAYS = 'Agreed response date cannot be more than 28 days after the orginal response date';
```
[the remark this answers is not in the document]
Comment (laperla11) [reply src/main/common/form/validationErrors/errorMessageConstants.ts]: removed it

File: src/main/services/features/response/responseDeadlineService.ts [hunk 4]
```diff
@@ -17,4 +19,36 @@ export class ResponseDeadlineService {
...
+    } catch (error) {
+      logger.error(error);
+      throw error;
+    }
+
+  }
+}
+
+function setDate(date: Date): AgreedResponseDeadline {
+  const dateOfResponseDeadline = new Date(date);
```
[the remark this answers is not in the document]
Comment (laperla11) [reply src/main/services/features/response/responseDeadlineService.ts]: removed it

File: src/main/views/features/response/responseDeadline/agreed-response-deadline.njk [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+{% from "../../../macro/contact-us-for-help.njk" import contactUsForHelp %}
+{% from "../../../macro/csrf.njk" import csrfProtection %}
+{% from "../../../macro/page-title.njk" import setPageTitle %}
+
+{% block pageTitle %}
+  {{ setPageTitle('PAGES.AGREED_RESPONSE_DEADLINE.PAGE_TITLE') }}
+{% endblock %}
+{% set hint = t('PAGES.ADMISSION_PAYMENT_DATE.FOR_EXAMPLE', {
+  nextDate: nextMonth.getDate(),
+  nextMonth: nextMonth.getMonth() + 1,
```
[the remark this answers is not in the document]
Comment (laperla11) [reply src/main/views/features/response/responseDeadline/agreed-response-deadline.njk]: I removed the logic to view

File: src/main/modules/i18n/locales/cy.json [hunk 6]
```diff
@@ -162,6 +163,11 @@
       "PAGE_TITLE": "Payment date",
       "TITLE": "Ar ba ddyddiad byddwch yn talu"
     },
+    "AGREED_RESPONSE_DEADLINE": {
+      "PAGE_TITLE": "Agree more time to respond",
```
[the remark this answers is not in the document]
Comment (laperla11) [reply src/main/modules/i18n/locales/cy.json:167]: done

File: src/main/common/form/validators/dateNotBeforeReferenceDate.ts [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+ */
+@ValidatorConstraint({name: 'dateNotMoreThanDurationValidator', async: false})
+export class DateNotBeforeReferenceDate implements ValidatorConstraintInterface {
+
+  validate(inputDate: Date, validationArguments?: ValidationArguments) {
+    const property = validationArguments.constraints[0];
+    const referenceDate = (validationArguments.object as any)[property];
+    const daysBetweenTwoDates = getNumberOfDaysBetweenTwoDays(inputDate, referenceDate);
+
+    if (inputDate !== null && daysBetweenTwoDates >= 0) {
```
[the remark this answers is not in the document]
Comment (laperla11) [reply src/main/common/form/validators/dateNotBeforeReferenceDate.ts]: done

File: src/main/common/form/validators/dateNotMoreThanDurationValidator.ts [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+@ValidatorConstraint({name: 'dateNotMoreThanDurationValidator', async: false})
+export class DateNotMoreThanDurationValidator implements ValidatorConstraintInterface {
+
+  validate(inputDate: Date, validationArguments?: ValidationArguments) {
+    const property = validationArguments.constraints[0];
+    const duration = validationArguments.constraints[1];
+    const referenceDate = (validationArguments.object as any)[property];
+    const daysBetweenTwoDates = getNumberOfDaysBetweenTwoDays(referenceDate, inputDate);
+
+    if (inputDate !== null && daysBetweenTwoDates > duration) {
```
[the remark this answers is not in the document]
Comment (laperla11) [reply src/main/common/form/validators/dateNotMoreThanDurationValidator.ts]: done

File: src/main/common/form/validators/dateNotBeforeReferenceDate.ts [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+      return false;
+    }
+    return true;
+  }
+
+  defaultMessage() {
+    return 'ERRORS.VALID_AGREED_RESPONSE_DATE_NOT_IN_THE_PAST';
```
[the remark this answers is not in the document]
Comment (laperla11) [reply src/main/common/form/validators/dateNotBeforeReferenceDate.ts:19]: this is the default error message if the decorator doesn't have error message, in this case the decorator has the same error message and it is covered with unit test.

File: src/main/routes/features/response/responseDeadline/agreedResponseDeadlineController.ts [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+    AGREED_T0_MORE_TIME_URL, async (req, res, next: express.NextFunction) => {
+      const originalResponseDeadline = claim?.respondent1ResponseDeadline;
+      const {year, month, day} = req.body;
+      const agreedResponseDeadlineDate = new AgreedResponseDeadline(year, month, day, originalResponseDeadline);
+      const form: GenericForm<AgreedResponseDeadline> = new GenericForm<AgreedResponseDeadline>(agreedResponseDeadlineDate);
+      await form.validate();
+
+      if (form.hasErrors()) {
+        res.render(agreedResponseDeadlineViewPath, {
+          form: form,
```
[the remark this answers is not in the document]
Comment (laperla11) [reply src/main/routes/features/response/responseDeadline/agreedResponseDeadlineController.ts]: done

File: src/main/routes/features/response/responseDeadline/agreedResponseDeadlineController.ts [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+agreedResponseDeadlineController
+  .get(
+    AGREED_T0_MORE_TIME_URL, async (req: express.Request, res: express.Response, next: express.NextFunction) => {
+      backLink = constructResponseUrlWithIdParams(req.params.id, RESPONSE_DEADLINE_OPTIONS_URL);
+      try {
+        claim = await getCaseDataFromStore(req.params.id);
+        const agreedResponseDeadline = responseDeadlineService.getAgreedResponseDeadline(claim);
+        res.render(agreedResponseDeadlineViewPath, {
+          form: new GenericForm(agreedResponseDeadline),
+          nextMonth: nextMonth,
```
[the remark this answers is not in the document]
Comment (laperla11) [reply src/main/routes/features/response/responseDeadline/agreedResponseDeadlineController.ts]: done

File: src/main/modules/i18n/locales/en.json [hunk 12]
```diff
@@ -162,6 +163,11 @@
       "PAGE_TITLE": "Payment date",
       "TITLE": "What date will you pay on?"
     },
+    "AGREED_RESPONSE_DEADLINE": {
+      "PAGE_TITLE": "Agree more time to respond",
```
[the remark this answers is not in the document]
Comment (laperla11) [reply src/main/modules/i18n/locales/en.json:167]: done

File: src/main/routes/features/response/responseDeadline/agreedResponseDeadlineController.ts [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+import {ResponseDeadlineService} from '../../../../services/features/response/responseDeadlineService';
+import {getCaseDataFromStore} from '../../../../modules/draft-store/draftStoreService';
+import {Claim} from '../../../../common/models/claim';
+
+const responseDeadlineService = new ResponseDeadlineService();
+const agreedResponseDeadlineViewPath = 'features/response/responseDeadline/agreed-response-deadline';
+const agreedResponseDeadlineController = express.Router();
+const nextMonth = new Date();
+nextMonth.setMonth(nextMonth.getMonth() + 1);
+let claim: Claim;
```
[the remark this answers is not in the document]
Comment (laperla11) [reply src/main/routes/features/response/responseDeadline/agreedResponseDeadlineController.ts]: I changed the logic to retrieve claim in both get and post request

File: src/main/views/features/response/responseDeadline/agreed-response-deadline.njk (hunk 1 above)
Comment (laperla11) [reply src/main/views/features/response/responseDeadline/agreed-response-deadline.njk]: changed to errorSummary macro

**Review by commitBlob — APPROVED** [review]

**Pull request merged**